### ISSUE TITLE
Combining files < 5mb + using regex

### ISSUE THREAD

**Issue by emptyr1** [xtream1101/s3-concat#1]
Hi great work here! Thanks for putting this :) 
I have a few questions:

- Possible to use regex in bucket names? I'm trying to combine ( as an example) ALL the json files in:
(let's say I'm trying to combine the following 3  files)
```
1. s3://fluentd/events/012345esdf/2017/12/a.json
2. s3://fluentd/events/9872absdd/2017/12/b.json
3. s3://fluentd/events/333333dd/2017/12/c.json
```

What should be my BUCKET string will be? 
Also I didn't understand what the difference between 
`out_path` argument and `path_to_concat` argument. 
FYI my files are < 5 MB so using `small_parts_threads=4`

**Comment by xtream1101** [comment]
It is not possible to use a regex in the folder name. This is a limitation of the way the the files are being listed https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.list_objects

I did just add a method `add_file` in version `0.1.1` where you can add single file at a time.
I updated the code snippet in the read me to show this. This way you can get the files you need and add as you want.

From the example you gave, your bucket name would be `fluentd`. That would mean that the first folder you see in the aws console is `events/`.

I also updated the variable in question `out_path` to `concatenated_file` to hopefully make it more clear. This is the file that the small files are saved into.

The variable `path_to_concat` is a _folder_ in s3 where the files are that need to be concatenated.

**Comment by emptyr1** [comment]
Awesome thanks @xtream1101 ! So this `add_file` method helps! :) Is there a way to concatenate files in different prefixes ( within same bucket). 
I have 100 json files in s3://events/prefixaa/....
and 100 json files in s3://events/prefixbb/...
Is it possible to concatenate all the files in these ^ 2 prefixes to create 1 json file? Or using `add_file` my only option here?

**Comment by xtream1101** [comment]
You can call `add_file()` or `add_files()` as much as you need. All files you add by the time you call `.concat()` will be joined together.

**Comment by emptyr1** [comment]
Great thanks! That worked! Lastly is it possible to do some sort of compression within the `MultipartUploadJob`.. The jobs seems to be using multi part upload with threading so I'm sure if that's possible? Can you point me to which I should look at? Compressing it before uploading OR changing it to something like parquet format?

**Comment by xtream1101** [comment]
You cannot compress or change formats in this process. You would need to do
that in a separate process after it is concatenated. This is a limitation
of how the multipart upload work as well as appending to files.

**Comment by emptyr1** [comment]
Thanks @xtream1101 .. I should probably create a new ticket for this but Another minor issue in the `MultipartUploadJob`. 
In the main constructor you have:
```
        if len(self.parts_list) > 0:
            self.upload_id = self._start_multipart_upload(s3)
            parts_mapping = self._assemble_parts(s3)
            self._complete_concatenation(s3, parts_mapping)

        elif len(self.parts_list) == 1:
            # can perform a simple S3 copy since there is just a single file
```
The `elif` will never get executed right? since 1 > 0 ?

**Comment by xtream1101** [comment]
Thanks for seeing that. Fixed in 0.1.2
